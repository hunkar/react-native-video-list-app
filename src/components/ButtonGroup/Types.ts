import {IconNode} from 'react-native-elements/dist/icons/Icon';
import {StyleProp, ViewStyle} from 'react-native';

//TODO Other styles can be customized. Can be added in the future.
/**
 * title: text of button
 * icon: icon of button
 * onPress: press event of button
 */
export type ButtonProps = {
  title: string;
  icon?: IconNode;
  onPress?: (() => void) | undefined;
};

/**
 * containerStyle: Styles for group container
 * buttons: array of ButtonProps
 */
export type ButtonGroupProps = {
  containerStyle?: StyleProp<ViewStyle>;
  buttons: ButtonProps[];
};
