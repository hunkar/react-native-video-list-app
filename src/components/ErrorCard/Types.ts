/**
 * title: Text of error.
 * onRetry: function that will be called on retry button clicked.
 */
export type ErrorCardProps = {
  title: string;
  onRetry?: () => void;
};
