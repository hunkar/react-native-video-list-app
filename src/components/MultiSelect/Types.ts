/**
 * id: id of item,
 * name: item name that will be shown
 * isSelected: you can set select status of item
 * value: extra value for item. For external use.
 */
export type MultiSelectItemProps = {
  id: string;
  name: string;
  isSelected?: boolean;
  value?: string | number | boolean;
};

/**
 * items: array of MultiSelectItemProps
 * selectedItems: id array of selected items
 * onChange: trigger function with selected items
 */
export type MultiSelectProps = {
  items: MultiSelectItemProps[];
  selectedItems?: string[];
  onChange?: (item: MultiSelectItemProps[]) => void;
};
