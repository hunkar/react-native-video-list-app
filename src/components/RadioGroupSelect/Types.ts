/**
 * id: item id
 * name: item name that will be show on the list
 * isSelected: item can be selectable with this prop
 * value: any value for external use.
 */
export type RadioGroupSelectItemProps = {
  id: string;
  name: string;
  isSelected?: boolean;
  value?: string | number | boolean;
};

/**
 * items: array of RadioGroupSelectItemProps items
 * selectedItem: id of selected item
 * onChange: trigger with selected item on change
 */
export type RadioGroupSelectProps = {
  items: RadioGroupSelectItemProps[];
  selectedItem?: string;
  onChange?: (item: RadioGroupSelectItemProps) => void;
};
