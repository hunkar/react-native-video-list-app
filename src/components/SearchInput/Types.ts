/**
 * placeholder: placeholder for searchBar
 * onChangeText: event for text change
 * changeTimeout: letter input frequency waiting range in ms 
 */
export type SearchInputProps = {
  placeholder?: string;
  onChangeText?: (text: string) => void;
  changeTimeout?: number;
};
