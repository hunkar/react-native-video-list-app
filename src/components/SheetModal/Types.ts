/**
 * title: title of modal
 * children: any ReactNode for modal child
 */
export type SheetModalProps = {
  title: string;
  children: React.ReactNode;
};
