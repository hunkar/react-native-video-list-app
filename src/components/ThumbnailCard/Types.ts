/**
 * id: Id of music.
 * title: Main title for singer name.
 * subTitle: Sub title for music name.
 * imageUrl: Url of video image.
 */
export type ThumbnailCardProps = {
  id: string;
  title: string;
  subTitle: string;
  imageUrl: string;
};
