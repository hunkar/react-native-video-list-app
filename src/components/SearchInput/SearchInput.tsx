import React from 'react';
import {Platform} from 'react-native';
import {SearchBar} from 'react-native-elements';
import {SearchBarBaseProps} from 'react-native-elements/dist/searchbar/SearchBar';
import {SearchInputProps} from './Types';

const SafeSearchBar = SearchBar as unknown as React.FC<SearchBarBaseProps>;

/**
 * Custom serach input. This component can provide delay for changing.
 * This can prevent high change frequency.
 * 
 * @param props SearchInputProps
 * @returns ReactNode
 */
const SearchInput = (props: SearchInputProps) => {
  const {
    placeholder = 'Type Here...',
    onChangeText,
    changeTimeout = 300,
  } = props;

  let changeSetTimeout: ReturnType<typeof setTimeout>;

  /**
   * This custom function provide delay for change event.
   * @param changedText string
   */
  const onChange = (changedText: string) => {
    clearTimeout(changeSetTimeout);

    changeSetTimeout = setTimeout(() => {
      if (onChangeText) {
        onChangeText(changedText);
      }

      clearTimeout(changeSetTimeout);
    }, changeTimeout);
  };

  return (
    <SafeSearchBar
      testID='searchBar'
      placeholder={placeholder}
      onChangeText={onChange}
      platform={Platform.OS === 'android' ? 'android' : 'ios'}
    />
  );
};

export default SearchInput;
