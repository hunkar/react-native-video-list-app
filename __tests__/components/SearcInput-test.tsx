import renderer, {act} from 'react-test-renderer';
import React from 'react';
import {SearchInputProps} from '../../src/components/SearchInput/Types';
import SearchInput from '../../src/components/SearchInput/SearchInput';

test('snapshot equility', async () => {
  const mockProps: SearchInputProps = {
    placeholder: 'Search',
    onChangeText: jest.fn(),
    changeTimeout: 500,
  };

  const component = renderer.create(<SearchInput {...mockProps} />);

  expect(component.toJSON()).toMatchSnapshot();
});

test('renders placeholder correct and platform should be ios', async () => {
  const mockProps: SearchInputProps = {
    placeholder: 'Search',
    onChangeText: jest.fn(),
    changeTimeout: 500,
  };

  const component = renderer.create(<SearchInput {...mockProps} />);

  const searchComponent = component.root.findByProps({
    testID: 'searchBar',
  }).props;

  expect(searchComponent.placeholder).toEqual(mockProps.placeholder);
  expect(searchComponent.platform).toEqual('ios');
});

test('onChange of searchbar should be change 3 times, but our onChangeText should be called once', done => {
  const mockProps: SearchInputProps = {
    placeholder: 'Search',
    onChangeText: jest.fn(),
    changeTimeout: 500,
  };

  const component = renderer.create(<SearchInput {...mockProps} />);

  const searchComponent = component.root.findByProps({
    testID: 'searchBar',
  }).props;

  act(() => {
    searchComponent.onChangeText('a');
    searchComponent.onChangeText('ab');
    searchComponent.onChangeText('abc');
  });

  setTimeout(() => {
    expect(mockProps.onChangeText).toBeCalledTimes(1);
    expect(mockProps.onChangeText).toBeCalledWith('abc');
    done();
  }, mockProps.changeTimeout || 0);
});

test('onChange of searchbar should be change 2 times with changeTimeout space, our onChangeText should be called 2 times', done => {
  const mockProps: SearchInputProps = {
    placeholder: 'Search',
    onChangeText: jest.fn(),
    changeTimeout: 200,
  };

  const component = renderer.create(<SearchInput {...mockProps} />);

  const searchComponent = component.root.findByProps({
    testID: 'searchBar',
  }).props;

  act(() => {
    searchComponent.onChangeText('a');
  });

  setTimeout(() => {
    act(() => {
      searchComponent.onChangeText('ab');
    });

    setTimeout(() => {
      expect(mockProps.onChangeText).toBeCalledTimes(2);
      done();
    }, mockProps.changeTimeout || 0);
  }, mockProps.changeTimeout || 0);
});
