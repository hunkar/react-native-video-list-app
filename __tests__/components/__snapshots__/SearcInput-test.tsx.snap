// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot equility 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "flexDirection": "row",
      "overflow": "hidden",
      "paddingBottom": 13,
      "paddingTop": 13,
    }
  }
>
  <View
    style={
      Object {
        "paddingHorizontal": 0,
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#dcdce1",
          "borderBottomWidth": 0,
          "borderColor": "#86939e",
          "borderRadius": 9,
          "flexDirection": "row",
          "marginLeft": 8,
          "marginRight": 8,
          "minHeight": 36,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginLeft": 8,
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#7d7d7d",
                      "fontSize": 20,
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                    Object {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
                testID="iconIcon"
              >
                ï”…
              </Text>
            </View>
          </View>
        </View>
      </View>
      <TextInput
        editable={true}
        onBlur={[Function]}
        onCancel={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Search"
        placeholderTextColor="#7d7d7d"
        platform="ios"
        style={
          Object {
            "color": "#242424",
            "flex": 1,
            "fontSize": 18,
            "marginLeft": 6,
            "minHeight": 40,
            "overflow": "hidden",
          }
        }
        testID="searchBar"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 40,
            "justifyContent": "center",
            "marginRight": 8,
            "marginVertical": 4,
            "paddingRight": 4,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "#ff190c",
          "fontSize": 12,
          "height": 0,
          "margin": 0,
          "padding": 0,
        }
      }
    />
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "opacity": 0,
        "position": "absolute",
        "right": null,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            undefined,
            undefined,
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007aff",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          Cancel
        </Text>
      </View>
    </View>
  </View>
</View>
`;
